### PR TITLE
Fix Deep Link for Statements page

### PR DESCRIPTION
**Pull request by tomivs**
### Details
On mobile devices, links to Statements page (new.expensify.com/statements/202202/) aren't opening in the app, instead, they open in the web browser.

### Fixed Issues
$ https://github.com/Expensify/App/issues/8185

### Tests
1. Open web browser
2. Click on links that start with "new.expensify.com/..."
3. Click on links to Statements page "new.expensify.com/statements/..."

NOTE: It's necessary to update the file "[new.expensify.com/apple-app-site-association](https://new.expensify.com/apple-app-site-association)" with [my version of the file](https://github.com/tomivs/App/tree/tomivs-deeplink/.well-known/apple-app-site-association) for all Deep Links to work on iOS.

Also, after iOS 14, Apple [uses its own CDN](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_associated-domains) to host a copy of the file so it may take a few hours to update. That's why I tested with my own domain.

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
Note: Deep Links don't work (and have never worked) on Android 12 and above because they are not implemented for those versions.

For iOS, Android (11 or lower).

- Login to NewDot on the app.
- From an email or external source, click on the link [https://new.expensify.com/statements/202202/](https://new.expensify.com/statements/202202/)
- Verify that the link is opened in the native app, and not the web browser.

Other platforms
Regression tests only

It's necessary to update the file "[new.expensify.com/apple-app-site-association](https://new.expensify.com/apple-app-site-association)" with [the new version of the file](https://github.com/Expensify/App/blob/main/.well-known/apple-app-site-association) for all Deep Links to work on iOS.

Also, after iOS 14, Apple [uses its own CDN](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_associated-domains) to host a copy of the file so it may take a few hours to update. That's why I tested with my own domain.

- [x] Verify that no errors appear in the JS console

#### Web
https://user-images.githubusercontent.com/6005972/164132794-541c7795-39a7-4cff-9770-0e5d07f5808c.mp4

#### iOS
https://user-images.githubusercontent.com/6005972/163654731-c29919cb-dba4-4c7f-b100-5ae0e413ddf3.mov

#### Android
https://user-images.githubusercontent.com/6005972/163653553-c881ccbe-cae2-41be-92da-4df9ff486efc.mp4

### CODE REVIEW

**Comment by tomivs** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by tomivs** [comment]
recheck

File: android/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -68,6 +68,7 @@
             <data android:scheme="https" android:host="new.expensify.com" android:pathPrefix="/bank-account"/>
             <data android:scheme="https" android:host="new.expensify.com" android:pathPrefix="/iou"/>
             <data android:scheme="https" android:host="new.expensify.com" android:pathPrefix="/enable-payments"/>
+            <data android:scheme="https" android:host="new.expensify.com" android:pathPrefix="/statements"/>
 
             <!-- Staging URLs -->
             <data android:scheme="https" android:host="staging.new.expensify.com" android:pathPrefix="/r"/>
```
Comment (rushatgabhane) [inline android/app/src/main/AndroidManifest.xml:74]: @tomivs let's add it for staging urls too?

**Comment by rushatgabhane** [comment]
@tomivs here are the steps that I followed to test this PR.
1. Build the app on a physical device (tried both dev and production)
2. Go to tomivs.com/linki
3. Click on new.expensify.com/statements/202202

But the link opens in a webpage. 
`npx uri-scheme open "https://new.expensify.com/statements/20202" --android` doesn't work either.


Am I doing something wrong? Could you please help me here. Thanks!



Platforms: Android 12 (Pixel 6), iOS 15.4 (iPhone 13).

https://user-images.githubusercontent.com/29673073/163881074-fb16ffad-7e2f-43b2-8f4e-663ac89595bf.mp4

**Comment by tomivs** [comment]
> Platform: Android 12 (Pixel 6).
@rushatgabhane it works on Android 11 or lower. We would have to change the SDK versions and update AndroidManifest, build.gradle(s), etc. That's possible but might break a few things.

If you check "main/android/build.gradle", the Android app is set up for `targetSdkVersion = 30` (Android 11).

I would recommend another pull request such as "Update Android SDK version to 31 (Android 11)" in order to avoid regression bugs.

But if you test with other Android versions, it should work.

**Comment by tomivs** [comment]
Since there were changes for [App Links on Android 12](https://developer.android.com/about/versions/12/behavior-changes-12#android-app-links-verification-changes), these would be the steps we would have to follow in order to make the App Links work on Android 12:
- Update `targetSdkVersion` to 31 (Android 12).
- Update some old _npm packages_ to versions that support Android 12.
- Update `AndroidManifest.xml` following Android 12 guidelines.
- Re-generate and update https://www.expensify.com/.well-known/assetlinks.json in order to [verify links](https://www.xda-developers.com/android-12-will-always-open-non-verified-links-in-the-default-browser/) for the New App (right now it's only set up for the old app).

Without following the steps mentioned above, the App Links won't work on Android 12.

What do you guys think? Should we go like this for now or update for Android 12? If so, will you create a new Upwork post?

Let me know. I can help with that too.

**Comment by rushatgabhane** [comment]
@tomivs thanks for sharing your insights! I think this is what we should do

1. Merge this PR.
2. You should log a new issue on `#expensify-open-source` slack channel, because you found the cause.
3. Submit a proposal when the new issue is created.

**Review by rushatgabhane — APPROVED** [review]
@thienlnam LGTM! 🎉
Please let us know if [these are the best steps to move forward](https://github.com/Expensify/App/pull/8656#issuecomment-1102879869).

<details>
<summary><h4> #### PR Reviewer Checklist</h4></summary>

- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

**Comment by rushatgabhane** [comment]
@tomivs I think the tests and QA steps could be simplified to this -

For iOS, Android and mWeb.
1. Login to NewDot.
2. Send this link to any user - `new.expensify.com/statements/202202`
4. Click on the link `new.expensify.com/statements/202202`.
6. Verify that the link is opened in the native app, and not the web browser.

**Comment by rushatgabhane** [comment]
@tomivs I think the tests and QA steps could be simplified to this -

For iOS, Android.

Login to NewDot.
Send this link to any user - new.expensify.com/statements/202202
Click on the link new.expensify.com/statements/202202.
Verify that the link is opened in the native app, and not the web browser.
For web and desktop
Regression tests only

**Comment by thienlnam** [comment]
We already have deep linking for reports, so it seems like from this comment https://github.com/Expensify/App/pull/8656#issuecomment-1102486931 those should also not be working on Android 12 or above?

As for next steps, it looks more like this:
1. Review / Update / Merge this PR
2. Have new.expensify.com/apple-app-site-association(https://new.expensify.com/apple-app-site-association) updated for this to be working in production
3. Test it on staging and production
4. And then as mentioned, You should log a new issue on #expensify-open-source slack channel
6. Submit a proposal when the new issue is created.

As for this PR, let's get some screenshots on all platforms - and will this work on desktop as well?

**Review by thienlnam — COMMENTED** [review]
We already have deep linking for reports, so it seems like from this comment https://github.com/Expensify/App/pull/8656#issuecomment-1102486931 those should also not be working on Android 12 or above?

As for next steps, it looks more like this:
1. Review / Update / Merge this PR
2. Have new.expensify.com/apple-app-site-association(https://new.expensify.com/apple-app-site-association) updated for this to be working in production
3. Test it on staging and production
4. And then as mentioned, You should log a new issue on #expensify-open-source slack channel
6. Submit a proposal when the new issue is created.

As for this PR, let's get some screenshots on all platforms - and will this work on desktop as well?

**Comment by rushatgabhane** [comment]
> We already have deep linking for reports, so it seems like from this comment https://github.com/Expensify/App/pull/8656#issuecomment-1102486931 those should also not be working on Android 12 or above?

That's right, deep linking for reports doesn't work on Android 12 and above.

> and will this work on desktop as well?

Yep, desktop works fine!

**Comment by tomivs** [comment]
> As for this PR, let's get some screenshots on all platforms - and will this work on desktop as well?

It works - as there are not regression issues -, however, if you mean opening web links in the desktop app - it doesn't work at all. For desktop, all links open in the browser. The reason is because Electron's API only allows handlers for [protocols](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app) such as `expensify://` (not HTTP URLs) and something like that hasn't been implemented for Expensify app. I could help out implementing it, but this PR is only for adding Statements page's Deep Link to the current implementation.

**Review by thienlnam — APPROVED** [review]

**Comment by rushatgabhane** [comment]
My bad, I confused opening links from the desktop app with deep linking.

https://user-images.githubusercontent.com/29673073/164271337-80f9aeca-085e-4b64-ae0c-be5603a57cb0.mov

**Comment by mvtglobally** [comment]
@tomivs @thienlnam are we able to QA this in staging? 
Team tried using https://staging.new.expensify.com/apple-app-site-association and it's not working. Not sure if this is a PROD PR only?

**Comment by thienlnam** [comment]
Yeah, I think this can only be tested on production for iOS due to their app association.  This would be a prod QA

**Comment by tomivs** [comment]
> @tomivs @thienlnam are we able to QA this in staging? Team tried using https://staging.new.expensify.com/apple-app-site-association and it's not working. Not sure if this is a PROD PR only?

It should be possible to test it both in staging and production, but the file needs to be updated in the website. I just checked the URL you shared and it's not updated. That might be the reason. If other URLs such as https://new.expensify.com/settings work, then the file `apple-app-association` needs to be updated in the website.

For staging: https://staging.new.expensify.com/apple-app-site-association
For production: https://new.expensify.com/apple-app-site-association

**Comment by kbecciv** [comment]
@thienlnam @tomivs QA team unable to test this PR on both  in staging and production.

**Comment by tomivs** [comment]
> @thienlnam @tomivs QA team unable to test this PR on both  in staging and production.

@kbecciv Files in the website must be updated...

For staging:
https://staging.new.expensify.com/apple-app-site-association

For production:
https://new.expensify.com/apple-app-site-association

**Comment by thienlnam** [comment]
Okay, I have some extra context on this.  The file should be updated automatically once it is deployed since you updated them.  The thing is, we're actually hosting an old file at those links so you won't see any of the new changes.  Apple has updated their specification to .well-known/apple-app-site-association located [here](https://developer.apple.com/documentation/xcode/supporting-associated-domains)
Screen Shot 2022-04-27 at 11 03 22 AM.  We've also found that we're deploying it to the wrong location due to webpack auto inferring the file type which is being fixed in https://github.com/Expensify/App/pull/8803

Next steps: Once that PR is merged and deployed, this PR can be tested properly

**Comment by mvtglobally** [comment]
@thienlnam Does it mean we should re-test this now?

**Comment by thienlnam** [comment]
@mvtglobally Yes, please retest this on staging and on production once it is there as well

**Comment by tomivs** [comment]
@kbecciv @mvtglobally Any update from QA team on this? @thienlnam mentioned a PR that needed to be merged and it has already been merged. So I believe it should be able to get tested on staging now.

**Comment by mvtglobally** [comment]
@tomivs sorry I should've posted here as well https://github.com/Expensify/App/issues/8777#issuecomment-1111685591
Retest is Pass

**Comment by tomivs** [comment]
How is QA on production going? I tested it out and it seems to be working.

cc: @kbecciv @mvtglobally

**Pull request merged**